### PR TITLE
Add NilCore library as a future replacement for DataTypes

### PR DESCRIPTION
**Pull request by aleasims**
Here instead of SSZ-container approach used in DataTypes we implemented `sszpp`-serialization directly on types.

This is how to use it:

```cpp
#include "ssz++.hpp"
#include "zkevm_framework/core/types/block.hpp"

int main() {
    std::vector<std::byte> bytes = read_ssz_file();
    core::types::Block block = ssz::deserialize<core::types::Block>(bytes);
    std::vector<std::byte> new_bytes = ssz::serialize(block);
    // assert(bytes == new_bytes)
}
```

This is just a first iteration of implementing Nil structures. Other structure implementations and their interfaces are coming.

### CODE REVIEW

File: tests/libs/nil_core/test_block.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    return bytes;
+}
+
+// This is a bad test which needs to be replaces with proper integration testing:
+// load ssz-serialized entity from nil and pass it to NilCore.
+
+TEST(NilCoreBlockTests, SerializeDeserializeBlock) {
+    const std::string hex_ssz =
```
Comment (akokoshn) [inline tests/libs/nil_core/test_block.cpp:26]: Let's avoid using predefined serialized values. It should be check on the separate integration tests.
I guess we can make easy support of unit test by using template function:
```
template <typename T>
test_serialization(const T& obj) {
    auto bytes = ssz::serialize(obj);
    auto new_obj = ssz::deserialize<T>(bytes);
    EXPECT_TRUE(compare<T>(new_obj, obj));
}
```

Comment (akokoshn) [reply tests/libs/nil_core/test_block.cpp:26]: and merge all tests into one file

Comment (aleasims) [reply tests/libs/nil_core/test_block.cpp:26]: Yeah, I added it just to check at least something before proper testing with `nil`. There is actually not much that we can test here in unit tests.

**Comment by aleasims** [comment]
Also added some helpers to read/write `core::Bytes` (may be used to read/write `core::types::Code` e.g.):

```cpp
core::types::Bytes<100> blob = {{std::byte{0xAA}, std::byte{0xBB}, std::byte{0xCC}}};
std::ofstream f("out.bin");
write_bytes(blob, f);
```

File: libs/nil_core/include/zkevm_framework/core/utils.hpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
+#ifndef ZKEMV_FRAMEWORK_LIBS_NIL_CORE_INCLUDE_ZKEVM_FRAMEWORK_CORE_UTILS_HPP_
```
Comment (akokoshn) [inline libs/nil_core/include/zkevm_framework/core/utils.hpp:1]: Why we actually need read/write_bytes?

**Comment by akokoshn** [comment]
> Also there are some helpers to read/write `core::Bytes` (may be used to read/write `core::types::Code` e.g.):
> 
> ```c++
> core::types::Bytes<100> blob = {{std::byte{0xAA}, std::byte{0xBB}, std::byte{0xCC}}};
> std::ofstream f("out.bin");
> write_bytes(blob, f);
> ```

I guess it's not so good idea - add extra functionality for this library.
It should be as easy as possible. Better locate helpers in some common utilities.
Actually we have to do these conversion only when parse JSON (file or response) - https://github.com/NilFoundation/zkEVM-framework/blob/ebc73a27fa1415cfae238a44ed322e50a67494fb/libs/json_helpers/json_helpers.cpp#L83

File: libs/nil_core/include/zkevm_framework/core/utils.hpp (hunk 2 above)
Comment (aleasims) [reply libs/nil_core/include/zkevm_framework/core/utils.hpp:1]: Well, in fact we don't. These functions were useful before when we had stream read/write interfaces. Now they may be useful if you will need to write `Bytes<N>`-based types into stream. 

But we can safely remove them, it's up to you, I just didn't want to remove anything which may be useful.

**Comment by aleasims** [comment]
@akokoshn sure, no problem, I will remove these functions.

**Review by akokoshn — APPROVED** [review]

**Pull request merged**